ext/geos_c_impl/factory.c

```diff
@@ -38,100 +38,175 @@ static void message_handler(const char* fmt, ...)
+
+  factory_data = (RGeo_FactoryData*)data;
+  if (!NIL_P(factory_data->wkrep_wkt_generator)) {
+    mark(factory_data->wkrep_wkt_generator);
   }
-  if (!NIL_P(data->wkrep_wkb_generator)) {
-    rb_gc_mark(data->wkrep_wkb_generator);
+  if (!NIL_P(factory_data->wkrep_wkb_generator)) {
+    mark(factory_data->wkrep_wkb_generator);
   }
-  if (!NIL_P(data->wkrep_wkt_parser)) {
-    rb_gc_mark(data->wkrep_wkt_parser);
+  if (!NIL_P(factory_data->wkrep_wkt_parser)) {
+    mark(factory_data->wkrep_wkt_parser);
   }
-  if (!NIL_P(data->wkrep_wkb_parser)) {
-    rb_gc_mark(data->wkrep_wkb_parser);
+  if (!NIL_P(factory_data->wkrep_wkb_parser)) {
+    mark(factory_data->wkrep_wkb_parser);
   }
-  if (!NIL_P(data->proj4_obj)) {
-    rb_gc_mark(data->proj4_obj);
+  if (!NIL_P(factory_data->proj4_obj)) {
+    mark(factory_data->proj4_obj);
   }
-  if (!NIL_P(data->coord_sys_obj)) {
-    rb_gc_mark(data->coord_sys_obj);
+  if (!NIL_P(factory_data->coord_sys_obj)) {
+    mark(factory_data->coord_sys_obj);
   }
 }
 
 
 // Mark function for geometry data. This marks the factory and klasses
 // held by the geometry so those don't get collected.
 
-static void mark_geometry_func(RGeo_GeometryData* data)
+static void mark_geometry_func(void* data)
+{
+  RGeo_GeometryData* geometry_data;
+
+  geometry_data = (RGeo_GeometryData*)data;
+  if (!NIL_P(geometry_data->factory)) {
+    mark(geometry_data->factory);
+  }
+  if (!NIL_P(geometry_data->klasses)) {
+    mark(geometry_data->klasses);
+  }
+}
+
+
+#ifdef HAVE_RB_GC_MARK_MOVABLE
+static void compact_factory_func(void* data)
 {
-  if (!NIL_P(data->factory)) {
-    rb_gc_mark(data->factory);
+  RGeo_FactoryData* factory_data;
+
+  factory_data = (RGeo_FactoryData*)data;
+  if (!NIL_P(factory_data->wkrep_wkt_generator)) {
+    factory_data->wkrep_wkt_generator = rb_gc_location(factory_data->wkrep_wkt_generator);
+  }
+  if (!NIL_P(factory_data->wkrep_wkb_generator)) {
+    factory_data->wkrep_wkb_generator = rb_gc_location(factory_data->wkrep_wkb_generator);
+  }
+  if (!NIL_P(factory_data->wkrep_wkt_parser)) {
+    factory_data->wkrep_wkt_parser = rb_gc_location(factory_data->wkrep_wkt_parser);
+  }
+  if (!NIL_P(factory_data->wkrep_wkb_parser)) {
+    factory_data->wkrep_wkb_parser = rb_gc_location(factory_data->wkrep_wkb_parser);
+  }
+  if (!NIL_P(factory_data->proj4_obj)) {
+    factory_data->proj4_obj = rb_gc_location(factory_data->proj4_obj);
   }
-  if (!NIL_P(data->klasses)) {
-    rb_gc_mark(data->klasses);
+  if (!NIL_P(factory_data->coord_sys_obj)) {
+    factory_data->coord_sys_obj = rb_gc_location(factory_data->coord_sys_obj);
   }
 }
 
 
+static void compact_geometry_func(void* data)
+{
+  RGeo_GeometryData* geometry_data;
+
+  geometry_data = (RGeo_GeometryData*)data;
+  if (!NIL_P(geometry_data->factory)) {
+    geometry_data->factory = rb_gc_location(geometry_data->factory);
+  }
+  if (!NIL_P(geometry_data->klasses)) {
+    geometry_data->klasses = rb_gc_location(geometry_data->klasses);
+  }
+}
```

Comment:
Since I'm not extremely familiar with all of the memory management internals could you explain how the `compact_` method is different than the `mark_` and why we need both?

I'm guessing compact is run before an object is being moved, but just want to clarify since they seem very similar.

Reply:
They're actually more or less two pieces of the same puzzle. We are using the movable variant `rb_gc_mark_movable` which tells the GC that this object can be moved. When the GC actually moves this object, it will then leave a trail in the form of a pointer, so the object does not get lost. As the result of this, we end up with pointers pointing other pointers. This is where `compact_` comes in, where we define the logic to remove the unnecessary pointers. @keithdoggett

Reply:
Gotcha that makes sense thanks for the explanation @teckwan